src/sites/dMCDA.jl

```diff
@@ -3,6 +3,33 @@
 using StatsBase
 using Distances
 using Combinatorics
+using JMcDM
+using ADRIA: order_ranking, adria_vikor, adria_topsis
+
+global method = [
```

Comment:
Globals are bad! Mark this a const for the module please.

Also, isn't there a way of getting a list of available methods from JMcDM?

Reply:
The list of available methods wasn't working for me when I tried previously. I just tried again and it's working now, so I'll add it.

Reply:
Actually, sorry I remember now why I used the global (which I can replace with a constant), it's because I needed to add something which indicates whether the method gives best sites as the highest score or lowest score (I add a Boolean next to the method) and I wasn't sure how to do this without manually looking at the docs for the method and working it out. I'll have a look again and see if there is something which indicates but otherwise not sure how to replace.

Reply:
So looks like the results struture also has a "best index" field, so I can add somethng which checks whether the best index is equal to the max or min score and order the sites accordingly :)

Reply:
So I think the reason the list of JMcDM functions wasn't working for me before is it uses the subtypes() function (in subtypes(MCDMMethod)) which is automatically loaded in interactive mode but doesn't seem to be loaded when running ADRIA. Due to this it throws an error saying it can't find the subtypes() function. Is it ok to include the InteractiveUtils module in dMCDA or is this not a good idea? If so, is there a way to just load one function from the module?

Reply:
No, don't add `InteractiveUtils` to our list of dependencies.

This should be fixed in JMcDM.

Could you give me a link to where this is being used in JMcDM please?

Reply:
This is where it's used- under the mcdm subtitle https://jbytecode.github.io/JMcDM/dev/utility/

It works fine when I use it in the command prompt, but when I brekpoint in dMCDA in ADRIA with "using JMcDM " in the file, it throws an error for subtypes() for me.

Reply:
Hi Rose, spoke to the maintainer of JMcDM and he's of the opinion that use of `subtypes()` was for illustrative purposes only and not a supported core feature of the package.

This means we're forced to add InteractiveUtils as a dependency.
After adding it to the project, please also add the below somewhere, either in the main `ADRIA.jl` file or in the file where it is used. I think the latter is preferable given we don't use it anywhere else.

```
using InteractiveUtils: subtypes
```

Reply:
Hi Takuya, thanks for checking up on this. Maybe if we add JMcDM as an extension later then we will only have to add the dependency if the extension is used. For now I'll add to dependecies as suggested.

Reply:
Hi Takuya, so I've been working on adding this now and it seems like the type of the elements of `subtypes(MCDMMethod)` are 'DataType` so they aren't actually functions and cause an error if you try and run the JMcDM methods with them. Do you know a way of getting the associated function from the datatype?

Reply:
Shouldn't matter (beauty of multiple dispatch). Calling the `DataType` (e.g., `SomeDataType()`) should resolve to the corresponding constructor.

If this is an issue with defining an argument type for a function, then I suggest to just use `Union{Function, DataType}`.

Reply:
Incidentally, if your implementations of TOPSIS/VIKOR/Order Ranking are equivalent to those in JMcDM, then I highly encourage you to submit a PR there to replace their slower implementations. Then we don't need to maintain functionally equivalent code.

Reply:
> Shouldn't matter (beauty of multiple dispatch). Calling the `DataType` (e.g., `SomeDataType()`) should resolve to the corresponding constructor.
> 
> If this is an issue with defining an argument type for a function, then I suggest to just use `Union{Function, DataType}`.

It's fine going into the function (I changed the input to DataType), but it errors when used with the JMcDM mcda() function, it complains it's a DataType not a JMcDM mcda function. I couldn't find anything on the JMcDM docs, they don't actually use the list to implement a method in the example. I'm looking at it again this morning, so I'll look at the docs again and see if I can find anything.

Reply:
> Incidentally, if your implementations of TOPSIS/VIKOR/Order Ranking are equivalent to those in JMcDM, then I highly encourage you to submit a PR there to replace their slower implementations. Then we don't need to maintain functionally equivalent code in ADRIA.

I would have to make changes because my functions don't allow for combinations of Max and min in the criteria but I can definitely do this at some point.

Reply:
When the JMcDM function is used in this line :
https://github.com/open-AIMS/ADRIA.jl/blob/b119fa91911edd4a957cfa859197df82f1219aed/src/sites/dMCDA.jl#L216

It should be something like `mcda_func=VikorMethod()`, but the entries from subtypes(JMcDM.MCDMMethod) error.

Reply:
Note, I've also tried `methods(mcda_func)[1]` to extract the method but because it's a general method and the mcda function requires a `JMcDM.MCDMMethod` type it errors.

Reply:
https://github.com/open-AIMS/ADRIA.jl/blob/ea5762f5b1f050a1f031ef9a538acb3741237329/src/sites/dMCDA.jl#L222C10-L229

That's what you want.

Note that the Promethee method errors because the constructor for it expects additional arguments.
There may be others like it.

Not sure what you want to do with these methods.

Also note the docstring is now out of date.

```julia
ERROR: MethodError: no method matching JMcDM.PROMETHEE.PrometheeMethod()

Closest candidates are:
  JMcDM.PROMETHEE.PrometheeMethod(::Vector{<:Function}, ::Vector{Float64}, ::Vector{Float64})
   @ JMcDM C:\Users\takuy\.julia\packages\JMcDM\Oo88a\src\promethee.jl:22                                                                                                                                                                                                                 |  ETA: 0:00:04        
  JMcDM.PROMETHEE.PrometheeMethod(::Vector{Function}, ::Vector{Union{Nothing, Float64}}, ::Vector{Union{Nothing, Float64}})
   @ JMcDM C:\Users\takuy\.julia\packages\JMcDM\Oo88a\src\promethee.jl:13
  JMcDM.PROMETHEE.PrometheeMethod(::Any, ::Any, ::Any)
   @ JMcDM C:\Users\takuy\.julia\packages\JMcDM\Oo88a\src\promethee.jl:13
```

Reply:
Ah man, brackets is all it needed :( ...
Promethee should be in the list of excluded methods because of this- it's more complicated because you get 2 scores are output and based on these you decide whether to have a single or multiple "best choice". I'll add it to the list. The others which I've added should be the only methods which are exceptions, I looked through all th docs to catch any which have non-standard outputs.

I'll update the docstring.

Reply:
I did say to call the constructor...

https://github.com/open-AIMS/ADRIA.jl/pull/348#discussion_r1201298621

Reply:
I didn't realise that's what you meant by your example, I thought you were just referring to the datatype of the inputs because you referred to multile dispatch